File: zimbra_tag_delta.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from __future__ import print_function

import argparse, json, os, re, subprocess, sys
from datetime import datetime
from pathlib import Path

TOOL_VERSION = "1.2.1"

# Default branch candidates (keep master/main out unless you pass them)
DEFAULT_BRANCHES = [
    "origin/release/{line}",
    "origin/support/{line}",
    "origin/hotfix/{line}",
]

# Heuristics
SECURITY_SUBJECT_RE = re.compile(
    r"(CVE-|security|vuln|sanitize|escape|validate|xss|csrf|xxe|rce|ssrf|sqli|auth|"
    r"login|session|cookie|token|password|privilege|dos|overflow|jetty|proxy|nginx|tls|ssl)",
    re.IGNORECASE,
)
SECURITY_FILE_RE = re.compile(
    r"(auth|login|session|cookie|csrf|xss|escape|sanitize|imap|smtp|lmtp|sieve|parser|upload|"
    r"attachment|mime|proxy|jetty|nginx|ssl|tls|crypto|sasl|clamav|spamassassin|jackson|netty|guava|log4j|openssl)",
    re.IGNORECASE,
)
CODE_EXT_RE = re.compile(r"\.(java|jsp|js|ts|c|cc|cpp|go|rb|py|xml|groovy|scala|kt)$", re.IGNORECASE)

MAX_COMMITS_ANALYZED = 800
NUMS_RE = re.compile(r"\d+")

# ---------------- subprocess helper (Py3.6 safe) ----------------
def run(cmd, cwd=None):
    p = subprocess.run(
        cmd, cwd=cwd,
        stdout=subprocess.PIPE, stderr=subprocess.PIPE,
        universal_newlines=True, check=False
    )
    return (p.stdout or "").strip(), (p.stderr or "").strip(), p.returncode

# ---------------- version/tag helpers ----------------
def parse_version_tuple(vstr):
    m = re.match(r"^\s*(\d+)\.(\d+)\.(\d+)\s*$", vstr)
    if not m:
        raise ValueError("must be major.minor.patch (e.g., 10.0.16)")
    return (int(m.group(1)), int(m.group(2)), int(m.group(3)))

def line_of(tup):
    return "{}.{}".format(tup[0], tup[1])

def tag_nums_for_line(tag, major, minor):
    nums = [int(n) for n in NUMS_RE.findall(tag)]
    if len(nums) < 2: return None
    if nums[0] != major or nums[1] != minor: return None
    while len(nums) < 3: nums.append(0)
    return tuple(nums[:3])

def nearest_le_tag(tags, major, minor, ceiling_tuple):
    best = None; best_nums = None
    for t in tags:
        nums = tag_nums_for_line(t, major, minor)
        if not nums: continue
        if nums <= ceiling_tuple:
            if best_nums is None or nums > best_nums:
                best, best_nums = t, nums
    return best, best_nums

def list_tags(repo_dir):
    out, _, _ = run(["git", "tag", "--list"], cwd=repo_dir)
    tags = [t.strip() for t in out.splitlines() if t.strip()]
    def key(t):
        n = [int(x) for x in NUMS_RE.findall(t)]
        return tuple(n)
    return sorted(tags, key=key)

def ref_exists(repo_dir, ref):
    _, _, rc = run(["git", "rev-parse", "--verify", "--quiet", ref], cwd=repo_dir)
    return rc == 0

def pick_branch(repo_dir, patterns, version_line):
    for pat in patterns:
        ref = pat.format(line=version_line)
        if ref_exists(repo_dir, ref): return ref
    return None

# ---------------- git ops ----------------
def fetch_refs(repo_dir, verbose=False):
    run(["git", "fetch", "--all", "--prune"], cwd=repo_dir)
    run(["git", "fetch", "--tags", "--force"], cwd=repo_dir)
    if verbose: print("  [fetched tags/branches]")

def rev_parse(repo_dir, ref):
    out, _, _ = run(["git", "rev-parse", ref], cwd=repo_dir)
    return out.strip()

def commits_between(repo_dir, base_ref, target_ref):
    out, _, _ = run(["git", "log", "--no-merges", "--pretty=%h\t%s", base_ref+".."+target_ref, "--"], cwd=repo_dir)
    return [ln for ln in out.splitlines() if ln.strip()]

def files_touched(repo_dir, full_sha):
    out, _, _ = run(["git", "show", "--name-only", "--pretty=format:", full_sha], cwd=repo_dir)
    return [f for f in out.splitlines() if f.strip()]

def commit_score(subject, files):
    score = 0
    if SECURITY_SUBJECT_RE.search(subject): score += 3
    hit_file = False
    for f in files:
        if SECURITY_FILE_RE.search(f): score += 2; hit_file = True
        if CODE_EXT_RE.search(f): score += 1
    if hit_file and score < 3: score += 1
    return score

# ---------------- main ----------------
def main():
    ap = argparse.ArgumentParser(description="Zimbra tag delta: nearest ≤ base tag → capped target (tag ≤ ceiling or branch).")
    ap.add_argument("--version", required=True, help="Base version (major.minor.patch), e.g., 10.0.15")
    ap.add_argument("--ceiling-tag", help="Ceiling (major.minor.patch). Default: base+1 patch")
    ap.add_argument("--ceiling-mode", choices=["skip","branch"], required=True,
                    help="If no tag ≤ ceiling exists after base: skip or fall back to branch tip")
    ap.add_argument("--repos-file", required=True, help="File with repo names (one per line)")
    ap.add_argument("--workdir", default=".", help="Directory holding repos (default: .)")
    ap.add_argument("--branches", nargs="*", default=DEFAULT_BRANCHES,
                    help="Branch fallback patterns; {line} placeholder allowed")
    ap.add_argument("--out", default="./tag_delta_out", help="Output directory (default: ./tag_delta_out)")
    ap.add_argument("--format", choices=["csv","md"], default="md", help="Also write a Markdown rollup (default md)")
    ap.add_argument("--debug", action="store_true", help="Verbose per-repo reasoning & commit list")
    ap.add_argument("-V","--tool-version", action="store_true", help="Print tool version and exit")
    args = ap.parse_args()

    if args.tool_version:
        print("zimbra_tag_delta.py", TOOL_VERSION); sys.exit(0)

    try:
        base_tuple = parse_version_tuple(args.version)
    except ValueError as e:
        print("Error parsing --version:", e, file=sys.stderr); sys.exit(2)

    if args.ceiling_tag:
        try:
            ceil_tuple = parse_version_tuple(args.ceiling_tag)
        except ValueError as e:
            print("Error parsing --ceiling-tag:", e, file=sys.stderr); sys.exit(2)
    else:
        ceil_tuple = (base_tuple[0], base_tuple[1], base_tuple[2] + 1)

    version_line = line_of(base_tuple)

    # load repos
    with open(args.repos_file, "r", encoding="utf-8") as f:
        repos = [ln.strip().split("/")[-1] for ln in f if ln.strip() and not ln.strip().startswith("#")]
    # de-dupe preserve order
    seen=set(); repos = [r for r in repos if not (r in seen or seen.add(r))]

    outdir = Path(args.out); outdir.mkdir(parents=True, exist_ok=True)
    ts = datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
    line_tag = version_line

    summary_rows = []
    manifest_rows = []
    details = {}

    changed_repos = 0
    analyzed_repos = 0

    for repo in repos:
        repo_dir = Path(args.workdir) / repo
        if args.debug: print("\n=== {} ===".format(repo))
        if not repo_dir.exists():
            if args.debug: print("SKIP: not cloned at {}".format(repo_dir))
            continue

        analyzed_repos += 1
        fetch_refs(repo_dir, verbose=args.debug)

        tags = list_tags(repo_dir)
        if args.debug: print("All tags ({}): {}".format(len(tags), ", ".join(tags)))

        base_tag, base_nums = nearest_le_tag(tags, base_tuple[0], base_tuple[1], base_tuple)
        if args.debug: print("Base tag (<= {}.{}.{}): {}".format(*base_tuple, base_tag))
        if not base_tag:
            if args.debug: print("SKIP: no base tag for line {}".format(version_line))
            continue

        target_ref = None
        target_is_tag = False

        tgt_tag, tgt_nums = nearest_le_tag(tags, base_tuple[0], base_tuple[1], ceil_tuple)
        if tgt_tag:
            target_ref = tgt_tag; target_is_tag = True
            if args.debug: print("Target tag (<= {}.{}.{}): {}".format(*ceil_tuple, target_ref))
        else:
            if args.ceiling_mode == "branch":
                b = pick_branch(repo_dir, args.branches, version_line)
                if b:
                    target_ref = b; target_is_tag = False
                    if args.debug: print("Target branch fallback:", target_ref)
            if not target_ref:
                if args.debug: print("SKIP: no target tag ≤ ceiling and ceiling-mode=skip")
                continue

        base_sha = rev_parse(repo_dir, base_tag)
        target_sha = rev_parse(repo_dir, target_ref)
        if base_sha == target_sha:
            if args.debug: print("SKIP: base == target ({} == {})".format(base_tag, target_ref))
            continue

        commits = commits_between(repo_dir, base_tag, target_ref)
        if args.debug:
            if commits:
                print("FOUND: {} commits between {}..{}".format(len(commits), base_tag, target_ref))
                for ln in commits[:40]: print("  "+ln)
                if len(commits) > 40: print("  ... ({} more)".format(len(commits)-40))
            else:
                print("No commits between {}..{}".format(base_tag, target_ref))

        suspicious_count = 0
        scored = []
        for ln in commits[:MAX_COMMITS_ANALYZED]:
            try:
                short, subject = ln.split("\t", 1)
            except ValueError:
                short, subject = ln, ""
            full_sha = rev_parse(repo_dir, short) or short
            files = files_touched(repo_dir, full_sha)
            sc = commit_score(subject, files)
            if sc > 0: suspicious_count += 1
            scored.append({"sha": full_sha, "short": short, "subject": subject, "score": sc, "files": files[:200]})

        top3 = sorted(scored, key=lambda x: x["score"], reverse=True)[:3]

        summary_rows.append({
            "repo": repo, "base_tag": base_tag, "base_sha": base_sha,
            "target_ref": target_ref, "target_sha": target_sha,
            "commit_count": len(commits), "suspicious_commits": suspicious_count,
            "top": [{"short": t["short"], "score": t["score"], "subject": t["subject"]} for t in top3],
        })
        manifest_rows.append({
            "repo": repo, "base_ref": base_tag, "base_sha": base_sha,
            "target_ref": target_ref, "target_sha": target_sha,
        })
        details[repo] = {
            "base": {"tag": base_tag, "sha": base_sha},
            "target": {"ref": target_ref, "sha": target_sha, "is_tag": target_is_tag},
            "commits_analyzed": len(scored),
            "suspicious_commits": suspicious_count,
            "commits": scored,
        }

        changed_repos += 1
        # ALWAYS print a concise line when not in --debug (so it never looks like “immediate return”)
        if not args.debug:
            print("[{}] {} → {} | commits: {} | suspicious: {}".format(
                repo, base_tag, target_ref, len(commits), suspicious_count
            ))

    # Sort & write outputs even if nothing changed
    summary_rows.sort(key=lambda r: (r["suspicious_commits"], r["commit_count"]), reverse=True)

    ts = datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
    summary_csv = outdir / ("tag_delta_summary_{}_{}.csv".format(line_of(base_tuple), ts))
    manifest_csv = outdir / ("tag_delta_manifest_{}_{}.csv".format(line_of(base_tuple), ts))
    details_json = outdir / ("tag_delta_details_{}_{}.json".format(line_of(base_tuple), ts))
    md_path = outdir / ("tag_delta_report_{}_{}.md".format(line_of(base_tuple), ts)) if args.format == "md" else None

    with summary_csv.open("w", encoding="utf-8") as f:
        f.write("repo,base_tag,base_sha,target_ref,target_sha,commit_count,suspicious_commits\n")
        for r in summary_rows:
            f.write("{repo},{base_tag},{base_sha},{target_ref},{target_sha},{commit_count},{suspicious_commits}\n".format(**r))

    with manifest_csv.open("w", encoding="utf-8") as f:
        f.write("repo,base_ref,base_sha,target_ref,target_sha\n")
        for r in manifest_rows:
            f.write("{repo},{base_ref},{base_sha},{target_ref},{target_sha}\n".format(**r))

    with details_json.open("w", encoding="utf-8") as f:
        json.dump(details, f, indent=2)

    if md_path:
        with md_path.open("w", encoding="utf-8") as f:
            f.write("# Zimbra Tag Delta — {} — {}\n\n".format(line_tag, ts))
            f.write("| Repo | Base → Target | Commits | Suspicious | Top suspects |\n|---|---|---:|---:|---|\n")
            for r in summary_rows:
                tops = r.get("top", [])
                top_txt = "<br/>".join("`{}` [{}] {}".format(t["short"], t["score"], t["subject"]) for t in tops) if tops else ""
                f.write("| `{}` | `{} → {}` | {} | {} | {} |\n".format(
                    r["repo"], r["base_tag"], r["target_ref"], r["commit_count"], r["suspicious_commits"], top_txt
                ))
            f.write("\n<sub>See CSV/JSON for exact SHAs and per-commit file lists.</sub>\n")

    # Always print a final summary
    print("\nAnalyzed repos: {} | Repos with changes: {}".format(analyzed_repos, changed_repos))
    print("Wrote:\n  {}\n  {}\n  {}".format(summary_csv, manifest_csv, details_json))
    if md_path: print("  {}".format(md_path))

if __name__ == "__main__":
    main()

File: zimbra_tag_delta.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Getopt::Long qw(GetOptions);
use POSIX qw(strftime);
use File::Spec;
use File::Path qw(make_path);
use JSON::PP;
use Cwd qw(abs_path);

# ---------------- Config / Heuristics ----------------
my $TOOL_VERSION = "1.2.1-perl";

my @DEFAULT_BRANCHES = (
  "origin/release/{line}",
  "origin/support/{line}",
  "origin/hotfix/{line}",
);

my $SECURITY_SUBJECT_RE = qr/(CVE-|security|vuln|sanitize|escape|validate|xss|csrf|xxe|rce|ssrf|sqli|auth|login|session|cookie|token|password|privilege|dos|overflow|jetty|proxy|nginx|tls|ssl)/i;
my $SECURITY_FILE_RE    = qr/(auth|login|session|cookie|csrf|xss|escape|sanitize|imap|smtp|lmtp|sieve|parser|upload|attachment|mime|proxy|jetty|nginx|ssl|tls|crypto|sasl|clamav|spamassassin|jackson|netty|guava|log4j|openssl)/i;
my $CODE_EXT_RE         = qr/\.(java|jsp|js|ts|c|cc|cpp|go|rb|py|xml|groovy|scala|kt)$/i;

my $MAX_COMMITS_ANALYZED = 800;

# ---------------- CLI ----------------
my ($version, $ceiling_tag, $ceiling_mode, $repos_file, $workdir, @branches, $outdir, $format, $debug, $print_tool_version);
$workdir = ".";
$outdir  = "./tag_delta_out";
$format  = "md";

GetOptions(
  "version=s"      => \$version,
  "ceiling-tag=s"  => \$ceiling_tag,
  "ceiling-mode=s" => \$ceiling_mode,  # skip | branch
  "repos-file=s"   => \$repos_file,
  "workdir=s"      => \$workdir,
  "branches=s@"    => \@branches,      # allow multiple
  "out=s"          => \$outdir,
  "format=s"       => \$format,        # csv | md
  "debug"          => \$debug,
  "tool-version|V" => \$print_tool_version,
) or die "Error parsing options. Try --help\n";

if ($print_tool_version) {
  print "zimbra_tag_delta.pl $TOOL_VERSION\n";
  exit 0;
}

sub usage {
  print <<"USAGE";
Usage:
  $0 --version <X.Y.Z> --ceiling-mode {skip|branch} --repos-file repos.txt [options]

Required:
  --version X.Y.Z          Base version (e.g., 10.0.15)
  --ceiling-mode MODE      'skip' (tag-to-tag only) or 'branch' (fallback to release branch tip)

Common:
  --ceiling-tag X.Y.Z      Upper bound tag (default: base + 1 patch)
  --repos-file FILE        List of repo dirs under --workdir (one per line)
  --workdir DIR            Root dir that already contains the repos (default: .)
  --branches PATTERN...    Branch fallback patterns (default: release/support/hotfix line)
  --out DIR                Output dir (default: ./tag_delta_out)
  --format md|csv          Also write Markdown rollup if 'md' (CSV/JSON always written)
  --debug                  Verbose per-repo reasoning
  -V, --tool-version       Print tool version and exit
USAGE
  exit 2;
}

usage() unless defined $version && defined $ceiling_mode && defined $repos_file;
die "--ceiling-mode must be 'skip' or 'branch'\n" unless $ceiling_mode eq 'skip' || $ceiling_mode eq 'branch';

@branches = @DEFAULT_BRANCHES unless @branches && @branches > 0;

# ---------------- Helpers ----------------
sub parse_version_tuple {
  my ($vstr) = @_;
  die "--version/--ceiling-tag must be major.minor.patch (e.g., 10.0.16)\n"
    unless $vstr =~ /^\s*(\d+)\.(\d+)\.(\d+)\s*$/;
  return ($1+0, $2+0, $3+0);
}

my ($base_major, $base_minor, $base_patch) = parse_version_tuple($version);
my ($ceil_major, $ceil_minor, $ceil_patch) =
  defined $ceiling_tag ? parse_version_tuple($ceiling_tag) : ($base_major, $base_minor, $base_patch + 1);

my $version_line = "$base_major.$base_minor";

sub run_git {
  my ($repo_dir, @cmd) = @_;
  my $old = Cwd::getcwd();
  chdir $repo_dir or return ("", "chdir failed", 1);
  my $cmd_str = join(" ", @cmd);
  my $out = qx{$cmd_str 2>&1};
  my $rc = $? >> 8;
  chdir $old;
  my $err = ($rc != 0) ? $out : "";
  $out = ($rc == 0) ? $out : "";
  chomp $out;
  return ($out, $err, $rc);
}

sub fetch_refs {
  my ($repo_dir) = @_;
  run_git($repo_dir, "git fetch --all --prune >/dev/null");
  run_git($repo_dir, "git fetch --tags --force >/dev/null");
  print "  [fetched tags/branches]\n" if $debug;
}

sub list_tags {
  my ($repo_dir) = @_;
  my ($out, $err, $rc) = run_git($repo_dir, "git tag --list");
  return [] if $rc != 0;
  my @tags = grep { $_ ne "" } split /\n/, $out;

  # Numeric-ish sort by extracting numbers
  my @sorted = sort {
    my @an = ($a =~ /(\d+)/g);
    my @bn = ($b =~ /(\d+)/g);
    my $i = 0;
    while ($i <= $#an || $i <= $#bn) {
      my $av = $i <= $#an ? $an[$i] : -1;
      my $bv = $i <= $#bn ? $bn[$i] : -1;
      return $av <=> $bv if $av != $bv;
      $i++;
    }
    return lc($a) cmp lc($b);
  } @tags;

  return \@sorted;
}

sub tag_nums_for_line {
  my ($tag, $maj, $min) = @_;
  my @nums = ($tag =~ /(\d+)/g);
  return undef if @nums < 2;
  return undef if $nums[0] != $maj || $nums[1] != $min;
  push @nums, 0 while @nums < 3;
  return [ @nums[0..2] ];
}

sub le_tuple {
  my ($a, $b) = @_; # arrays [maj,min,patch]
  return 1 if ($a->[0] < $b->[0]) || ($a->[0] == $b->[0] && $a->[1] <  $b->[1]) ||
              ($a->[0] == $b->[0] && $a->[1] == $b->[1] && $a->[2] <= $b->[2]);
  return 0;
}

sub tuple_cmp {
  my ($a, $b) = @_;
  return ($a->[0] <=> $b->[0]) || ($a->[1] <=> $b->[1]) || ($a->[2] <=> $b->[2]);
}

sub nearest_le_tag {
  my ($tags, $maj, $min, $ceil_tuple) = @_;
  my $best; my $best_nums;
  foreach my $t (@$tags) {
    my $nums = tag_nums_for_line($t, $maj, $min) or next;
    if (le_tuple($nums, $ceil_tuple)) {
      if (!defined $best_nums || tuple_cmp($nums, $best_nums) > 0) {
        $best = $t; $best_nums = $nums;
      }
    }
  }
  return ($best, $best_nums);
}

sub ref_exists {
  my ($repo_dir, $ref) = @_;
  my ($out, $err, $rc) = run_git($repo_dir, "git rev-parse --verify --quiet $ref");
  return $rc == 0;
}

sub pick_branch {
  my ($repo_dir, $patterns, $line) = @_;
  foreach my $pat (@$patterns) {
    my $ref = $pat; $ref =~ s/\{line\}/$line/g;
    return $ref if ref_exists($repo_dir, $ref);
  }
  return undef;
}

sub rev_parse {
  my ($repo_dir, $ref) = @_;
  my ($out, $err, $rc) = run_git($repo_dir, "git rev-parse $ref");
  return $rc == 0 ? $out : "";
}

sub commits_between {
  my ($repo_dir, $base_ref, $target_ref) = @_;
  my $cmd = "git log --no-merges --pretty=format:%H%x00%s $base_ref..$target_ref --";
  my ($out, $err, $rc) = run_git($repo_dir, $cmd);
  return [] if $rc != 0;
  my @lines = grep { $_ ne "" } split /\n/, $out;
  return \@lines;
}

sub files_touched {
  my ($repo_dir, $sha) = @_;
  my ($out, $err, $rc) = run_git($repo_dir, "git show --name-only --pretty=format: $sha");
  return [] if $rc != 0;
  my @files = grep { $_ ne "" } split /\n/, $out;
  return \@files;
}

sub commit_score {
  my ($subject, $files) = @_;
  my $score = 0;
  $score += 3 if $subject =~ $SECURITY_SUBJECT_RE;
  my $hit_file = 0;
  foreach my $f (@$files) {
    if ($f =~ $SECURITY_FILE_RE) { $score += 2; $hit_file = 1; }
    if ($f =~ $CODE_EXT_RE)      { $score += 1; }
  }
  $score += 1 if $hit_file && $score < 3;
  return $score;
}

# ---------------- Load repos ----------------
open my $RF, "<", $repos_file or die "Cannot open --repos-file $repos_file: $!\n";
my @repos;
my %seen;
while (my $ln = <$RF>) {
  chomp $ln;
  next if $ln =~ /^\s*$/ || $ln =~ /^\s*#/;
  $ln =~ s/.*\///; # keep last path part
  next if $seen{$ln}++;
  push @repos, $ln;
}
close $RF;

# ---------------- Outputs ----------------
make_path($outdir);
my $ts = strftime("%Y%m%dT%H%M%SZ", gmtime());
my $line_tag = "$base_major.$base_minor";

my $summary_csv  = File::Spec->catfile($outdir, "tag_delta_summary_${line_tag}_$ts.csv");
my $manifest_csv = File::Spec->catfile($outdir, "tag_delta_manifest_${line_tag}_$ts.csv");
my $details_json = File::Spec->catfile($outdir, "tag_delta_details_${line_tag}_$ts.json");
my $md_path      = $format eq 'md' ? File::Spec->catfile($outdir, "tag_delta_report_${line_tag}_$ts.md") : undef;

open my $SUM, ">", $summary_csv or die "Write $summary_csv: $!";
print $SUM "repo,base_tag,base_sha,target_ref,target_sha,commit_count,suspicious_commits\n";

open my $MAN, ">", $manifest_csv or die "Write $manifest_csv: $!";
print $MAN "repo,base_ref,base_sha,target_ref,target_sha\n";

my %details;
my @summary_rows; # for MD sorting
my $analyzed_repos = 0;
my $changed_repos  = 0;

# ---------------- Main loop ----------------
foreach my $repo (@repos) {
  my $repo_dir = File::Spec->catdir($workdir, $repo);
  print "\n=== $repo ===\n" if $debug;
  unless (-d $repo_dir) {
    print "SKIP: not cloned at $repo_dir\n" if $debug;
    next;
  }

  $analyzed_repos++;
  fetch_refs($repo_dir) if $debug; # always fetch, but print only in debug
  run_git($repo_dir, "git fetch --tags --force >/dev/null"); # ensure tags even if not debug

  my $tags = list_tags($repo_dir);
  print "All tags (@$tags)\n" if $debug && @$tags;

  my ($base_tag, $base_nums) = nearest_le_tag($tags, $base_major, $base_minor, [$base_major,$base_minor,$base_patch]);
  print "Base tag (<= $base_major.$base_minor.$base_patch): " . (defined $base_tag ? $base_tag : "(none)") . "\n" if $debug;
  next unless defined $base_tag;

  my $target_ref;
  my $target_is_tag = 0;
  my ($tgt_tag, $tgt_nums) = nearest_le_tag($tags, $base_major, $base_minor, [$ceil_major,$ceil_minor,$ceil_patch]);
  if (defined $tgt_tag) {
    $target_ref = $tgt_tag; $target_is_tag = 1;
    print "Target tag (<= $ceil_major.$ceil_minor.$ceil_patch): $target_ref\n" if $debug;
  } else {
    if ($ceiling_mode eq 'branch') {
      my $picked = pick_branch($repo_dir, \@branches, $version_line);
      if ($picked) {
        $target_ref = $picked; $target_is_tag = 0;
        print "Target branch fallback: $target_ref\n" if $debug;
      }
    }
    unless ($target_ref) {
      print "SKIP: no target tag ≤ ceiling and ceiling-mode=skip\n" if $debug;
      next;
    }
  }

  my $base_sha   = rev_parse($repo_dir, $base_tag);
  my $target_sha = rev_parse($repo_dir, $target_ref);

  if ($base_sha eq $target_sha) {
    print "SKIP: base == target ($base_tag == $target_ref)\n" if $debug;
    next;
  }

  my $commits = commits_between($repo_dir, $base_tag, $target_ref);
  if ($debug) {
    if (@$commits) {
      print "FOUND: " . scalar(@$commits) . " commits between $base_tag..$target_ref\n";
      my $show = 0;
      foreach my $ln (@$commits) {
        print "  $ln\n";
        last if (++$show >= 40);
      }
      print "  ... (" . (scalar(@$commits) - 40) . " more)\n" if @$commits > 40;
    } else {
      print "No commits between $base_tag..$target_ref\n";
    }
  }

  my $suspicious_count = 0;
  my @scored;
  my $count = 0;
  foreach my $ln (@$commits) {
    last if ++$count > $MAX_COMMITS_ANALYZED;
    my ($full_sha, $subject) = split /\0/, $ln, 2;   # NUL-delimited
    $subject = "" unless defined $subject;
    my $short = substr($full_sha, 0, 9);             # derive short from full
    my $files = files_touched($repo_dir, $full_sha);
    my $score = commit_score($subject, $files);
    $suspicious_count++ if $score > 0;
    push @scored, {
      sha     => $full_sha,
      short   => $short,
      subject => $subject,
      score   => $score,
      files   => [ @$files[0..($#$files < 199 ? $#$files : 199)] ],
    };
  }

  # top 3
  my @top = sort { $b->{score} <=> $a->{score} } @scored;
  splice(@top, 3) if @top > 3;

  # Write row aggregates
  print $SUM join(",", map { _csv($_) } ($repo,$base_tag,$base_sha,$target_ref,$target_sha, scalar(@$commits), $suspicious_count))."\n";
  print $MAN join(",", map { _csv($_) } ($repo,$base_tag,$base_sha,$target_ref,$target_sha))."\n";

  $details{$repo} = {
    base => { tag => $base_tag, sha => $base_sha },
    target => { ref => $target_ref, sha => $target_sha, is_tag => $target_is_tag ? JSON::PP::true : JSON::PP::false },
    commits_analyzed => scalar(@scored),
    suspicious_commits => $suspicious_count,
    commits => \@scored,
  };

  push @summary_rows, {
    repo => $repo, base_tag => $base_tag, base_sha => $base_sha,
    target_ref => $target_ref, target_sha => $target_sha,
    commit_count => scalar(@$commits),
    suspicious_commits => $suspicious_count,
    top => [ map { { short => $_->{short}, score => $_->{score}, subject => $_->{subject} } } @top ],
  };

  $changed_repos++;

  # concise line when not debug
  if (!$debug) {
    print "[$repo] $base_tag → $target_ref | commits: ".scalar(@$commits)." | suspicious: $suspicious_count\n";
  }
}

close $SUM;
close $MAN;

# Sort for MD
@summary_rows = sort {
  ($b->{suspicious_commits} <=> $a->{suspicious_commits}) || ($b->{commit_count} <=> $a->{commit_count})
} @summary_rows;

# details JSON
open my $DJ, ">", $details_json or die "Write $details_json: $!";
print $DJ JSON::PP->new->ascii->pretty->canonical->encode(\%details);
close $DJ;

# markdown
if (defined $md_path && $format eq 'md') {
  open my $MD, ">", $md_path or die "Write $md_path: $!";
  print $MD "# Zimbra Tag Delta — $version_line — $ts\n\n";
  print $MD "| Repo | Base → Target | Commits | Suspicious | Top suspects |\n|---|---|---:|---:|---|\n";
  foreach my $r (@summary_rows) {
    my $top_txt = "";
    if (@{$r->{top}}) {
      my @lines = map { sprintf("`%s` [%d] %s", $_->{short}, $_->{score}, _md_escape($_->{subject})) } @{$r->{top}};
      $top_txt = join("<br/>", @lines);
    }
    printf $MD "| `%s` | `%s → %s` | %d | %d | %s |\n",
      $r->{repo}, $r->{base_tag}, $r->{target_ref}, $r->{commit_count}, $r->{suspicious_commits}, $top_txt;
  }
  print $MD "\n<sub>See CSV/JSON for exact SHAs and per-commit file lists.</sub>\n";
  close $MD;
}

# summary to stdout
print "\nAnalyzed repos: $analyzed_repos | Repos with changes: $changed_repos\n";
print "Wrote:\n  $summary_csv\n  $manifest_csv\n  $details_json\n";
print "  $md_path\n" if defined $md_path;

exit 0;

# ---------------- utils ----------------
sub _csv {
  my ($v) = @_;
  $v = "" unless defined $v;
  if ($v =~ /[",\n]/) {
    $v =~ s/"/""/g;
    return "\"$v\"";
  }
  return $v;
}

sub _md_escape {
  my ($s) = @_;
  $s = "" unless defined $s;
  $s =~ s/\|/\\|/g;
  return $s;
}

